networkx/drawing/layout.py

```diff
@@ -1110,6 +1111,108 @@ def multipartite_layout(G, subset_key="subset", align="vertical", scale=1, cente
+):
+    """Arf layout for networkx
+
+    The  attractive   and  repulsive  forces   (arf)  layout (Geipel 2006)
+    improves  the spring  layout  in three  ways. First,  it
+    prevents  congestion of  highly connected  nodes due  to
+    strong forcing  between nodes.  Second, it  utilizes the
+    layout space  more effectively by preventing  large gaps
+    that  spring layout  tends  to create.  Lastly, the  arf
+    layout represents symmmetries in  the layout better than
+    the default spring layout.
+
+    Parameters
+    ----------
+    g : nx.Graph or nx.DiGraph
+        Networkx graph.
+    pos : dict
+        Initial  position of  the nodes.  If set  to None  a
+        random layout will be used using seed as input.
+    b : float
+        Controls the radius of  the circular space the graph
+        layout will be plotted in.
+    alpha : float
+        Attraction scalar between edges. Should be larger than 1.
+    etol : float
+        Termination tolerance
+    dt : float
+        Integration factor
+    max_iter : int
+        Max iterations before termination of the algorithm.
+    seed : int, RandomState instance or None  optional (default=None)
+        Set the random state for deterministic node layouts.
+        If int, `seed` is the seed used by the random number generator,
+        if numpy.random.RandomState instance, `seed` is the random
+        number generator,
+        if None, the random number generator is the RandomState instance used
+        by numpy.random.
+
+    References
+    - https://www.worldscientific.com/doi/abs/10.1142/S0129183107011558
+
+    Returns
+    -------
+    pos : dict
+        A dictionary of positions keyed by node.
+
+    Examples
+    --------
+    >>> G = nx.grid_graph((5, 5))
+    >>> pos = nx.arf_layout(G)
+
+    """
+    import numpy as np
+
+    assert alpha > 1, "Alpha should be larger than 1"
+    pos_tmp = nx.random_layout(g, seed=seed)
+    if pos is None:
+        pos = pos_tmp.copy()
+    else:
+        for node in g.nodes():
+            if node not in pos:
+                pos[node] = pos_tmp[node].copy()
+
+    # Initialize spring constant matrix
+    N = len(g)
+    # No nodes no computation
+    if N == 0:
+        return pos
+    K = np.zeros((N, N))
+    for idx, x in enumerate(g.nodes()):
+        for jdx, y in enumerate(g.nodes()):
+            if idx == jdx:
+                K[idx, jdx] = 0
+            elif g.has_edge(x, y) or g.has_edge(y, x):
+                K[idx, jdx] = alpha
+            else:
+                K[idx, jdx] = 1
+    # vectorize values
+    p = np.asarray(list(pos.values()))
+    bV = np.sqrt(N) * b  # scaling constant for circular space
+
+    # looping variables
+    error = etol + 1
+    n_iter = 0
+    while error > etol:
+        diff = p[:, None] - p[None]
+        A = np.linalg.norm(diff, axis=-1)[..., None]
+        # attraction_force - repulsions force
+        change = K[..., None] * diff - bV / A * diff
```

Comment:
I'm not sure what best practices are these days. But at one point it was deemed more readable to use `np.newaxis` instead of `None`. Thoughts?         Similarly, the ellipsis might be easier to read if it is written out as `K[:, :, None]`. But these are just questions, not suggestions.

In the last line, why not factor out the `diff` multiplication. I think the paper does it that way, and it is fewer multiplications, but is there another reason to keep the multiplications separate?

Reply:
I commonly use `None` due to it being shorter than `np.newaxis` but for clarity and interface compatibility I think it is better to use `np.newaxis` (although in practice I still end up using `None` a lot, behind the scenes it is the same anyway). I prefer the ellipses as it is clear that the last axis is used.